scipy/optimize/_linprog_rs.py

```diff
@@ -0,0 +1,535 @@
+
+    if header:
+        print("Phase",
+              "Iteration",
+              "Minimum Slack      ",
+              "Constraint Residual",
+              "Objective          ")
+
+    # :<X.Y left aligns Y digits in X digit spaces
+    fmt = '{0:<6}{1:<10}{2:<20.13}{3:<20.13}{4:<20.13}'
+    try:
+        slack = np.min(slack)
+    except ValueError:
+        slack = "NA"
+    print(fmt.format(phase, iteration, slack, np.linalg.norm(con), fun))
+
+
+def _phase_two(c, A, x, b, maxiter, tol, maxupdate, mast, pivot, iteration=0,
+               callback=None, _T_o=[], disp=False, phase_one_n=None):
+    """
+    The heart of the simplex method. Beginning with a basic feasible solution,
+    moves to adjacent basic feasible solutions successively lower reduced cost.
+    Terminates when there are no basic feasible solutions with lower reduced
+    cost or if the problem is determined to be unbounded.
+
+    This implementation follows the revised simplex method based on LU
+    decomposition. Rather than maintaining a tableau or an inverse of the
+    basis matrix, we keep a factorization of the basis matrix that allows
+    efficient solution of linear systems while avoiding stability issues
+    associated with inverted matrices.
+    """
+    m, n = A.shape
+    status = 0
+    a = np.arange(n)                    # indices of columns of A
+    ab = np.arange(m)                   # indices of columns of B
+    if maxupdate:
+        # basis matrix factorization object; similar to B = A[:, b]
+        B = BGLU(A, b, maxupdate, mast)
+    else:
+        B = LU(A, b)
+
+    for iteration in range(iteration, iteration + maxiter):
+
+        if disp or callback is not None:
+            if phase_one_n is not None:
+                phase = 1
+                x_postsolve = x[:phase_one_n]
+            else:
+                phase = 2
+                x_postsolve = x
+            x_o, fun, slack, con, _, _ = _postsolve(x_postsolve, *_T_o,
+                                                    tol=tol)
+
+            if callback is not None:
+                res = OptimizeResult({'x': x_o, 'fun': fun, 'slack': slack,
+                                      'con': con, 'nit': iteration,
+                                      'phase': phase, 'complete': False,
+                                      'status': 0, 'message': "",
+                                      'success': False})
+                callback(res)
+            else:
+                _display_iter(phase, iteration, slack, con, fun)
+
+        bl = np.zeros(len(a), dtype=bool)
+        bl[b] = 1
+
+        xb = x[b]       # basic variables
+        cb = c[b]       # basic costs
+
+        try:
+            v = B.solve(cb, transposed=True)    # similar to v = solve(B.T, cb)
+        except LinAlgError:
+            status = 4
+            break
+
+        # TODO: cythonize?
+        c_hat = c - v.dot(A)    # reduced cost
+        c_hat = c_hat[~bl]
+        # Above is much faster than:
```

Comment:
Is the commented out code supposed to stay after the merge?

Reply:
I intended to leave that comment there, yes, because it describes what we should be doing here. At some point it would be nice to write Cython code that does the same thing but more efficiently.